### ISSUE TITLE
[BUG] `avg_method` and `symmetric_cbar` flags using `plot_surf_stat_map` working properly with plotly engine

### ISSUE THREAD

**Issue by albertofpena** [nilearn/nilearn#4296]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Operating system

- [X] Linux
- [ ] Mac
- [ ] Windows

### Operating system version

Debian testing (trixie)

### Python version

- [ ] 3.12
- [X] 3.11
- [ ] 3.10
- [ ] 3.9
- [ ] 3.8

### nilearn version

0.10.3

### Expected behavior

When using the plotly engine, the colorbar should not be symmetric if symmetric_cbar=False. The avg_method should be user settable. This seems related to #3941 bug.

### Current behavior & error messages

This is what I got:

The flag symmetric_cbar=False is ignored, always plotting symmetric colorbars. On the other hand, when avg_method flag is inserted, I get the following error:

```bash
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[7], line 3
      1 from nilearn import plotting
----> 3 fig = plotting.plot_surf_stat_map(
      4     surf, texture, hemi='left',
      5     title='Surface left hemisphere', colorbar=True,
      6     engine='plotly', symmetric_cbar=False, avg_method='mean',
      7     darkness=0.5, cmap='black_red'
      8 )
     10 # change plotly lighting
     11 fig.figure.data[0].lighting.ambient = 0.5

File ~/.local/lib/python3.11/site-packages/nilearn/plotting/surf_plotting.py:1151, in plot_surf_stat_map(surf_mesh, stat_map, bg_map, hemi, view, engine, threshold, alpha, vmin, vmax, cmap, colorbar, symmetric_cbar, cbar_tick_format, bg_on_data, darkness, title, title_font_size, output_file, axes, figure, **kwargs)
   1142 # Call get_colorbar_and_data_ranges to derive symmetric vmin, vmax
   1143 # And colorbar limits depending on symmetric_cbar settings
   1144 cbar_vmin, cbar_vmax, vmin, vmax = get_colorbar_and_data_ranges(
   1145     loaded_stat_map,
   1146     vmin=vmin,
   1147     vmax=vmax,
   1148     symmetric_cbar=symmetric_cbar,
   1149 )
-> 1151 display = plot_surf(
   1152     surf_mesh, surf_map=loaded_stat_map,
   1153     bg_map=bg_map, hemi=hemi,
   1154     view=view, engine=engine, avg_method='mean', threshold=threshold,
   1155     cmap=cmap, symmetric_cmap=True, colorbar=colorbar,
   1156     cbar_tick_format=cbar_tick_format, alpha=alpha,
   1157     bg_on_data=bg_on_data, darkness=darkness,
   1158     vmax=vmax, vmin=vmin,
   1159     title=title, title_font_size=title_font_size, output_file=output_file,
   1160     axes=axes, figure=figure, cbar_vmin=cbar_vmin,
   1161     cbar_vmax=cbar_vmax, **kwargs
   1162 )
   1163 return display

TypeError: nilearn.plotting.surf_plotting.plot_surf() got multiple values for keyword argument 'avg_method'


```

### Steps and code to reproduce bug

```python

stat_img = image.smooth_img("TFCE_log_pDR_0002.nii", fwhm="fast")
surf = surface.load_surf_mesh('~/Downloads/Surf_Ice/BrainNet/BrainMesh_ICBM152_smoothed.lh.gii')
texture = surface.vol_to_surf(stat_img, surf, radius=3, kind='ball')

fig = plotting.plot_surf_stat_map(
    surf, texture, hemi='left',
    title='Surface left hemisphere', colorbar=True,
    engine='plotly', symmetric_cbar=False, avg_method='mean',
    darkness=0.5
)

# change plotly lighting
fig.figure.data[0].lighting.ambient = 0.5
fig.figure.data[0].lighting.diffuse = 0.6
fig.figure.data[0].lighting.specular = 0.1
fig.figure.data[0].lighting.roughness = 0.03
fig.figure.data[0].lighting.fresnel = 0.05
fig.figure.data[0].lightposition = dict(x=0, y=20, z=30)

fig.show()

```

**Comment by bthirion** [comment]
Thx for reporting. Could you make it work with images provided or generated by Nilearn  ? This would make it easier for us.

**Comment by Remi-Gau** [comment]
For the avg_method, the error you are getting is because `plot_surf_stat_map` will call `plot_surf` with `avg_method='mean'` so you cannot at the moment pass it another value (or even the same).

https://github.com/nilearn/nilearn/blob/1afacef948b6a3887530c3e683b1fc1e6930c733/nilearn/plotting/surf_plotting.py#L1158

Also note that avg_method is not implemented for plotly:

https://nilearn.github.io/stable/modules/generated/nilearn.plotting.plot_surf.html#nilearn.plotting.plot_surf

![image](https://github.com/nilearn/nilearn/assets/6961185/9275048a-31dd-4e91-9e15-ec4fa0292b13)

**Comment by Remi-Gau** [comment]
Went down the git blame 🐰 hole and it seems that this `avg_method='mean'` has been there for a long time.

This is the PR that set it I think.
https://github.com/nilearn/nilearn/pull/1387

@bthirion: any reason why we should force users to only plot the mean for a surface?

**Comment by Remi-Gau** [comment]
I can reproduce the symmetric_bar issue with plotly with this code.

```python
import numpy as np

from nilearn import datasets, plotting, surface

import matplotlib.pyplot as plt

stat_img = datasets.load_sample_motor_activation_image()

fsaverage = datasets.fetch_surf_fsaverage()

curv_right = surface.load_surf_data(fsaverage.curv_right)
curv_right_sign = np.sign(curv_right)

texture = surface.vol_to_surf(stat_img, fsaverage.pial_right)

engine = "matplotlib"

plotting.plot_surf_stat_map(
    fsaverage.infl_right,
    texture,
    hemi="right",
    title="Surface right hemisphere",
    colorbar=True,
    threshold=1.0,
    bg_map=curv_right_sign,
    engine=engine,
    symmetric_cbar=False,
)
plt.show()

engine = "plotly"

print(f"Using plotting engine {engine}.")

fig = plotting.plot_surf_stat_map(
    fsaverage.infl_right,
    texture,
    hemi="right",
    title="Surface right hemisphere",
    colorbar=True,
    threshold=1.0,
    bg_map=curv_right_sign,
    bg_on_data=True,
    engine=engine,
    symmetric_cbar=False,
)
fig.show()
```

**Comment by Remi-Gau** [comment]
OK it seems that in #3941 we fixed using the symmetric cmap issue with plotly for plotting ROIs ... but not for the other surface plotting functions... 
# 🤦🏾

**Comment by Remi-Gau** [comment]
Will open a draft PR to discuss this...

**Comment by bthirion** [comment]
> Went down the git blame 🐰 hole and it seems that this `avg_method='mean'` has been there for a long time.
> 
> This is the PR that set it I think. #1387
> 
> @bthirion: any reason why we should force users to only plot the mean for a surface?

AFAIK there is no technical reason.

**Comment by Remi-Gau** [comment]
> quick check and removing that `avg_method='mean'` for plot_surf_stat_map and trying to pass min or max makes no difference from what I can tell... Which is... surprising...

probably explained by the fact that a statistical map is a 3D volume to there is only one value along the 4th dimension, so min, max, mean will always be the same.

**Comment by Remi-Gau** [comment]
> Probably it's my mistake, but I assumed that the average was done at a vertex/face level.

Gosh... You are completely correct...

**Comment by Remi-Gau** [comment]
> Also, in this case you pass the statistical map as a texture, not as a 3D volume per se, so I'm not sure how it would affect.

I was trying to stay close to the code you provided but this did lead me to try passing the path to a nifti image as stat map as `stat_map` as the doc string says it should be ok but this does not seem to work:


```python
import matplotlib.pyplot as plt
import numpy as np

from nilearn import datasets, plotting, surface

fsaverage = datasets.fetch_surf_fsaverage()
curv_left = surface.load_surf_data(fsaverage.curv_left)
curv_left_sign = np.sign(curv_left)

stat_map = datasets.load_sample_motor_activation_image()

print(stat_map)

# %%
engine = "matplotlib"

plotting.plot_surf_stat_map(
    fsaverage.infl_left,
    stat_map = stat_map,
    hemi="left",
    title="Surface left hemisphere",
    cmap = 'bwr',
    colorbar=True,
    threshold=1.0,
    bg_map=curv_left_sign,
    engine=engine,
    symmetric_cbar=False,
    avg_method="mean",
)
```

gives

```
Traceback (most recent call last):
  File "/home/remi/github/nilearn/nilearn/tmp.py", line 22, in <module>
    plotting.plot_surf_stat_map(
  File "/home/remi/github/nilearn/nilearn/nilearn/plotting/surf_plotting.py", line 1167, in plot_surf_stat_map
    display = plot_surf(
              ^^^^^^^^^^
  File "/home/remi/github/nilearn/nilearn/nilearn/plotting/surf_plotting.py", line 809, in plot_surf
    fig = _plot_surf_matplotlib(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi/github/nilearn/nilearn/nilearn/plotting/surf_plotting.py", line 580, in _plot_surf_matplotlib
    surf_map_faces = _compute_surf_map_faces_matplotlib(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi/github/nilearn/nilearn/nilearn/plotting/surf_plotting.py", line 366, in _compute_surf_map_faces_matplotlib
    surf_map_data = _check_surf_map(surf_map, n_vertices)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi/github/nilearn/nilearn/nilearn/plotting/surf_plotting.py", line 345, in _check_surf_map
    raise ValueError("'surf_map' can only have one dimension "
ValueError: 'surf_map' can only have one dimension but has '3' dimensions
```

**Comment by Remi-Gau** [comment]
OK so I will split this into several PR to solve because there is more than one issue:
- dealing with avg_method
- dealing with symmetric_cbar
- and this thing that nifti cannot be passed to surface plotting even though the doc says so.

**Comment by Remi-Gau** [comment]
For the `symmetric_cbar` with plot_surf_stat_map.

Just writing downs the things to check.

- data
  - only has positive data
  - only has negative data
  - has positive and negative data
- `threshold` parameter
  - positive value
  - negative value
  - none 
- `symmetric_cbar` parameter:
  - False
  - True
  - "auto" 
- vmin
- vmax

**Comment by Remi-Gau** [comment]
Note passing a negative values as threshold is ignored and does not throw warnings:

```python
"""Test plotting asymmetric color bar."""

import matplotlib.pyplot as plt
import numpy as np

from nilearn import datasets, plotting, surface
from nilearn.image import threshold_img

vmin = None
vmax = None
symmetric_cbar = False
threshold = -0.5 

stat_img = datasets.load_sample_motor_activation_image()

fsaverage = datasets.fetch_surf_fsaverage()

curv_right = surface.load_surf_data(fsaverage.curv_right)
curv_right_sign = np.sign(curv_right)

texture = surface.vol_to_surf(stat_img, fsaverage.pial_right)

engine = "matplotlib"

plotting.plot_surf_stat_map(
    fsaverage.infl_right,
    texture,
    hemi="right",
    title="Surface right hemisphere",
    colorbar=True,
    threshold=threshold,
    bg_map=curv_right_sign,
    engine=engine,
    symmetric_cbar=False,
    cmap="black_red",
    vmin = vmin,
    vmax = vmax
)
plt.show()
```

gives 

![Figure_1](https://github.com/nilearn/nilearn/assets/6961185/8a396465-354a-4202-8f1d-cf949b64bb10)

Where as I would expect the same as for `threshold = 0.5` given that the threshold is supposed to act on absolute value according to the doc.

![Figure_2](https://github.com/nilearn/nilearn/assets/6961185/91982f34-a8f4-4dfe-a46f-3e5e38e729a1)

**Comment by Remi-Gau** [comment]
Tried passing invalid values to `symmetric_cbar` and I not getting errors. Will add some input validation.

**Comment by bthirion** [comment]
It's unclear to me whether the threshold has "one-tail" or "two-tail" behavior. The doc implies that is it two-tail, but we cold have cases where a one-tail behavior is expected ?

**Comment by Remi-Gau** [comment]
> It's unclear to me whether the threshold has "one-tail" or "two-tail" behavior. The doc implies that is it two-tail, but we cold have cases where a one-tail behavior is expected ?

That was why I was testing this. Less common but visualizing "deactivations" only may be  something users want to do. 

I'd be in favor of doing this in another PR after #4338 (even it means reverting it) just to make sure that we first align doc and behavior with the minimum amount of changes to the code.

**Comment by Remi-Gau** [comment]
@hndgzkn 
maybe this could be a follow up thing, you could have a look at: I opened a PR ages ago but feel free to close it and start another one if you prefer

**Comment by hndgzkn** [comment]
@Remi-Gau I've looked at the code and see that plotly and matplotlib code exists in the same module and other modules as well. I would like to spend some time on the code to refactor a little bit to separate modules for plotly and matplotlib, and also extract common parts to avoid separate implementations which do the same thing. Is that ok with you? Doing this, I think I will also fix the bug.

**Comment by hndgzkn** [comment]
@Remi-Gau I think I can start dealing with the problems listed in this issue. I'll start with `symmetric_cbar` problem in plotly. 

#4313  seems to fix the problem. However instead of merging the PR, it is closed. Do you remember why?

**Comment by Remi-Gau** [comment]
> Do you remember why?

not sure

probably it did not exactly do what I expected

**Comment by Remi-Gau** [comment]
I am using this script on 0.12.0 and this seems that the symmetric_cbar is fixed: at least I am getting the plots that I expect

@hndgzkn can you confirm?


```python
import numpy as np

from nilearn import datasets, plotting, surface, image

import matplotlib.pyplot as plt

stat_img = datasets.load_sample_motor_activation_image()

tresholded_stat_img = image.threshold_img(stat_img, threshold=1, copy_header=True, two_sided=False)

fsaverage = datasets.fetch_surf_fsaverage()

curv_right = surface.load_surf_data(fsaverage.curv_right)
curv_right_sign = np.sign(curv_right)

engine = "matplotlib"

for comment, img in zip(["stat_img", "tresholded_stat_img"], [stat_img, tresholded_stat_img]):
    texture = surface.vol_to_surf(img, fsaverage.pial_right)   
    for symmetric_cbar in [True, False, "auto"]:
        plotting.plot_surf_stat_map(
            fsaverage.infl_right,
            texture,
            hemi="right",
            title=f"{comment} - {symmetric_cbar=}",
            colorbar=True,
            threshold=0,
            bg_map=curv_right_sign,
            engine=engine,
            symmetric_cbar=symmetric_cbar,
            darkness=None,
        )
plotting.show()

engine = "plotly"

print(f"Using plotting engine {engine}.")

for comment, img in zip(["stat_img", "tresholded_stat_img"], [stat_img, tresholded_stat_img]):
    texture = surface.vol_to_surf(img, fsaverage.pial_right)   
    for symmetric_cbar in [True, False, "auto"]:
        fig = plotting.plot_surf_stat_map(
            fsaverage.infl_right,
            texture,
            hemi="right",
            title=f"{comment} - {symmetric_cbar=}",
            colorbar=True,
            threshold=1,
            bg_map=curv_right_sign,
            bg_on_data=True,
            engine=engine,
            symmetric_cbar=symmetric_cbar,
            darkness=None,

        )
        fig.show()
```

**Comment by Remi-Gau** [comment]
I am getting some warnings but some of them make sense.

```
UserWarning: you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True
```

This one I am ok with but it could be made more informative: if people want to use a symmetric_cmap=False on images that contain both positive and negative images, they should first threshold the image.

```
UserWarning:

vmin cannot be chosen when cmap is symmetric
```

This one is more confusing as the script does not request any value for `vmin`, so we should make sure it is not thrown.

Sell all warnings:
```
Using plotting engine plotly.
/home/remi-gau/github/nilearn/nilearn/tmp/cbar_sym.py:42: UserWarning: you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True
  fig = plotting.plot_surf_stat_map(
/home/remi-gau/github/nilearn/nilearn/tmp/cbar_sym.py:42: UserWarning: vmin cannot be chosen when cmap is symmetric
  fig = plotting.plot_surf_stat_map(
/home/remi-gau/github/nilearn/nilearn/tmp/cbar_sym.py:42: UserWarning:

you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True

/home/remi-gau/github/nilearn/nilearn/tmp/cbar_sym.py:42: UserWarning:

vmin cannot be chosen when cmap is symmetric

/home/remi-gau/github/nilearn/nilearn/tmp/cbar_sym.py:42: UserWarning:

you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True

/home/remi-gau/github/nilearn/nilearn/tmp/cbar_sym.py:42: UserWarning:

vmin cannot be chosen when cmap is symmetric
```

**Comment by hndgzkn** [comment]
> I am using this script on 0.12.0 and this seems that the symmetric_cbar is fixed: at least I am getting the plots that I expect
> 
> [@hndgzkn](https://github.com/hndgzkn) can you confirm?
> 

It is a bit strange that with your code it seems to work fine. But with the code below, it is still symmetric. Could you try it too?

I couldn't figure out the reason.

```
import matplotlib.pyplot as plt
from nilearn import datasets, plotting, surface

import numpy as np

stat_img = datasets.load_sample_motor_activation_image()

fsaverage = datasets.fetch_surf_fsaverage()

curv_right = surface.load_surf_data(fsaverage.curv_right)
curv_right_sign = np.sign(curv_right)

texture = surface.vol_to_surf(stat_img, fsaverage.pial_right)

engine = "plotly"

print(f"Using plotting engine {engine}.")

fig = plotting.plot_surf_stat_map(
    fsaverage.infl_right,
    texture,
    hemi="right",
    title="Surface right hemisphere",
    colorbar=True,
    threshold=1.0,
    bg_map=curv_right_sign,
    engine=engine,
    symmetric_cbar=False,
)

fig.show()
```

**Comment by hndgzkn** [comment]
> I am getting some warnings but some of them make sense.
> 
> ```
> UserWarning: you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True
> ```
> 
> This one I am ok with but it could be made more informative: if people want to use a symmetric_cmap=False on images that contain both positive and negative images, they should first threshold the image.
> 

I think even without thresholding they should be able to set symmetric_cmap=False, non? Instead we can give a warning to tell that, it might be better to use symmetric_cmap=True, when there are negative values, or threshold it. But I might be ignorent about usage.


> ```
> UserWarning:
> 
> vmin cannot be chosen when cmap is symmetric
> ```
> 
> This one is more confusing as the script does not request any value for `vmin`, so we should make sure it is not thrown.
> 

This is due to `symmetric_cbar`, `symmetric_cmap` not existing as parameter for all functions. `plot_surf_stat_map` has this param, but not `plot_surf`. Inside  plot_surf_stat_map, `vmin` is calculated depending on the value of `symmetric_cbar` and passed to `plot_surf`. We can decide the expected behavior and modify the code depending on that. 

I think it would also be good to decide on the meaning, behavior and necessity of these two parameters.

**Comment by Remi-Gau** [comment]
> But with the code below, it is still symmetric. Could you try it too?
> 
> I couldn't figure out the reason.

you get this warning that explains the reason:

```
/home/remi-gau/github/nilearn/nilearn/tmp/cbar2.py:19: UserWarning: you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True
```

I was getting similar result in my script when the non thresholded data was plotted.

**Comment by Remi-Gau** [comment]
> I think even without thresholding they should be able to set symmetric_cmap=False, non? Instead we can give a warning to tell that, it might be better to use symmetric_cmap=True, when there are negative values, or threshold it.

yeah that's the other possibility:

- so either we agree that the current behavior (force symmetric_cmap=True if the data contains both positive and negative values) is fine but maybe the warning / documentation ("threshold data first before plotting") should be improved to reduce the surprise on the outcome.
- or we agree that symmetric_cmap=False should be effective whatever the data content

Also we probably want to make sure the behavior and warnings are the same for all our plotting backend.

**Comment by Remi-Gau** [comment]
> We can decide the expected behavior

at least I would say that we should not throw a warning when the user has not passed a value of vmin that's different from the default.

**Comment by hndgzkn** [comment]
> you get this warning that explains the reason:
> 
> ```
> /home/remi-gau/github/nilearn/nilearn/tmp/cbar2.py:19: UserWarning: you have specified symmetric_cmap=False but the map contains negative values; setting symmetric_cmap to True
> ```
> 
> I was getting similar result in my script when the non thresholded data was plotted.

I'm running the code in notebook. When I run your code, plotly opens browser tabs to view the plots.

The plot in new tab is:

![Image](https://github.com/user-attachments/assets/ba1bfa67-3cb6-4d7d-b65f-ce08d955910a)


It also displays the plot in notebook output as follows:

![Image](https://github.com/user-attachments/assets/45949b60-1ab2-4a22-9982-ac6cb434ef77)


If I do not run in the loop, the output is as follows:

![Image](https://github.com/user-attachments/assets/d4e67675-bcc3-4713-baf7-c4d0fbc4438c)

I am trying to understand the reason for the difference.

I would think that the second image is what it should display.

**Comment by hndgzkn** [comment]
Ok, I think I understand. If I plot both matplotlib and plotly, the plots that I see in the notebook output are the matplotlib ones. And plotly opens a new tab for each plot. I was not attentive enough for the difference.

But the problem is matplotlib and plotly has different behaviors. Matplotlib lets users to have unsymmetric cbar even if there are negative values, but plotly does not. We should decide which behavior will be common for both.

I find, changing symmetric_cbar=True, even if the user sets it to False a bit restrictive (and actually it is partially the reason for this issue) .

**Comment by Remi-Gau** [comment]
OK let's try to align the plotly behavior on the one by matplotlib then.

if we can we should have one PR for each issue: the symmetric_cbar issue and the vmin warning issue

@hndgzkn do you want to take care of it?

**Comment by bthirion** [comment]
>     * so either we agree that the current behavior (force symmetric_cmap=True if the data contains both positive and negative values) is fine but maybe the warning / documentation ("threshold data first before plotting") should be improved to reduce the surprise on the outcome.

I'm not in favor of forcing behavior against the user's inputs. We should follow the users instructions and minimize warnings.
